libs-rs/eth-client/src/lib.rs

```diff
@@ -248,63 +260,59 @@ impl EthBridge {
+            }
+            if header_number > self.finalized_gc_threshold {
+                self.gc_headers(header_number - self.finalized_gc_threshold);
+            }
         }
     }
 
-    /// Removes old headers beyond the finality.
-    fn maybe_gc(&mut self, last_best_number: u64, new_best_number: u64) {
-        if new_best_number > last_best_number && last_best_number >= NUMBER_OF_BLOCKS_FINALITY {
-            for number in last_best_number - NUMBER_OF_BLOCKS_FINALITY
-                ..new_best_number - NUMBER_OF_BLOCKS_FINALITY
-            {
-                if let Some(mut hashes) = self.recent_header_hashes.get(&number) {
-                    env::log(
-                        format!("Removing {} old header(s) at #{}", hashes.len(), number)
-                            .as_bytes(),
-                    );
-                    for hash in hashes.iter() {
-                        self.infos.remove(&hash);
-                        self.headers.remove(&hash);
-                    }
-                    hashes.clear();
-                    self.recent_header_hashes.remove(&number);
+    /// Remove hashes from the canonical chain that are at least as old as the given header number.
+    fn gc_canonical_chain(&mut self, mut header_number: u64) {
+        loop {
+            if self.canonical_header_hashes.get(&header_number).is_some() {
+                self.canonical_header_hashes.remove(&header_number);
+                if header_number == 0 {
+                    break;
+                } else {
+                    header_number -= 1;
                 }
+            } else {
+                break;
             }
-            env::log(
-                format!("There are {} headers remaining", self.headers.len()).as_bytes(),
-            );
         }
     }
 
-    fn add_recent_header_hash(&mut self, number: u64, hash: &H256) {
-        let mut hashes = self.recent_header_hashes.get(&number).unwrap_or_else(|| {
-            let mut set_id = Vec::with_capacity(9);
-            set_id.extend_from_slice(b"s");
-            set_id.extend(number.to_le_bytes().iter());
-            UnorderedSet::new(set_id)
-        });
-        hashes.insert(&hash);
-        self.recent_header_hashes.insert(&number, &hashes);
+    /// Remove information about the headers that are at least as old as the given header number.
+    fn gc_headers(&mut self, mut header_number: u64) {
+        loop {
+            if let Some(all_headers) = self.all_header_hashes.get(&header_number) {
+                for hash in all_headers {
+                    self.headers.remove(&hash);
+                    self.infos.remove(&hash);
+                }
+                self.all_header_hashes.remove(&header_number);
+                if header_number == 0 {
+                    break;
+                } else {
+                    header_number -= 1;
+                }
+            } else {
+                break;
```

Comment:
Is there any chance record_header fail, and then it add another header, and there's hole in gc?
```
header_number:10 header_number:11 (empty due to fail to record_header) header_number:12
```
Then before header_number:10 will never be cleaned

Reply:
Someone will have to resubmit it.